backend/eventd/entity.go

```diff
@@ -20,39 +24,86 @@ func createProxyEntity(event *corev2.Event, s store.EntityStore) error {
 	}
 
 	// Determine if the entity exists
-	ctx := context.WithValue(context.Background(), corev2.NamespaceKey, event.Entity.Namespace)
-	entity, err := s.GetEntityByName(ctx, entityName)
-	if err != nil {
-		return err
-	}
+	//NOTE(ccressent): there is no timeout for this operation?
+	entityMeta := corev2.NewObjectMeta(entityName, namespace)
 
-	// If the entity does not exist, create a proxy entity
-	if entity == nil {
-		if event.Check.ProxyEntityName != "" {
-			// Create a brand new entity since we can't rely on the provided entity,
-			// which represents the agent's entity
-			entity = &corev2.Entity{
-				EntityClass:   corev2.EntityProxyClass,
-				Subscriptions: []string{corev2.GetEntitySubscription(entityName)},
-				ObjectMeta: corev2.ObjectMeta{
-					Namespace: event.Entity.Namespace,
-					Name:      entityName,
-				},
-			}
-		} else {
-			// Use on the provided entity
-			entity = event.Entity
-			entity.EntityClass = corev2.EntityProxyClass
-			entity.Subscriptions = append(entity.Subscriptions, corev2.GetEntitySubscription(entityName))
+	state := corev3.NewEntityState(namespace, entityName)
+	config := corev3.NewEntityConfig(namespace, entityName)
+
+	configReq := storev2.NewResourceRequestFromResource(context.Background(), config)
+	stateReq := storev2.NewResourceRequestFromResource(context.Background(), state)
+
+	var wState *wrap.Wrapper
+	var wConfig *wrap.Wrapper
+
+	wConfig, err := s.Get(configReq)
+	if err == nil {
+		if err := wConfig.UnwrapInto(config); err != nil {
+			return err
+		}
+
+		// Since the entity config exists, we fetch its associated state in
+		// order to create a fully formed corev2.Entity for the event.
+		wState, err = s.Get(stateReq)
+		if err != nil {
+			return err
+		}
+
+		if err := wState.UnwrapInto(state); err != nil {
+			return err
 		}
+	} else if err != nil {
+		switch err.(type) {
+		case *store.ErrNotFound:
+			// If the entity does not exist, create a proxy entity
+			if event.Check.ProxyEntityName != "" {
+				// Create a brand new entity since we can't rely on the provided
+				// entity, which represents the agent's entity
+				state.SetMetadata(&entityMeta)
+				config.SetMetadata(&entityMeta)
+			} else {
+				// Use on the provided entity
+				config, state = corev3.V2EntityToV3(event.Entity)
+			}
+
+			state.Metadata.CreatedBy = event.CreatedBy
```

Comment:
Should we set LastSeen here?

Reply:
I think we could if we wanted to, but as far as I can tell, the original code doesn't do it. For that reason, and to have the best chance of not breaking something subtle/unforeseen, I'd rather stick to that, at least in this pass.

Also, setting/updating `LastSeen` is probably the responsibility of `keepalived` rather than `eventd`?

Reply:
Ah you're totally right, that shouldn't happen here.